### ISSUE TITLE
GIL lock in multithread environment

### ISSUE THREAD

**Issue by BenTalagan** [cowboyd/therubyracer#432]
It seems to me redundant with other issues, but not expressed like this specifically : when trying to run **therubyracer** in a thread, the whole MRI is locked until the thread times out or ends by itself.

```
2.3.1 :001 > require "therubyracer"
2.3.1 :002 > ctx = V8::Context.new timeout:10000; 
2.3.1 :003> Thread.new{ ctx.eval("while(true) {}"); puts "Done."} # => MRI is locked for 10 seconds
```

If I understand correctly, that issue should be solved by the [1.0 release](https://github.com/cowboyd/therubyracer/pull/348) right? (can't wait for it!)  

**mini_racer** has that feature, but it hasn't that great bridge **therubyracer** has, so for the moment there is no 'ideal' solution for having a strong JS interpreter in a ruby env for the MRI.

**Comment by cowboyd** [comment]
It is true that a lot of deadlock issues will be solved with a 1.0 release. However, I have stalled out a bit because I no longer use therubyracer daily.

I'm currently looking for somebody to help push it over the finish line.

**Comment by Skarlso** [comment]
@cowboyd What does it take to get that over the finish line? Are there any potential Collaborators around?

**Comment by cowboyd** [comment]
There are probably about 20-40 hours of work left, and I am ready, happy and willing to pair with anybody willing to learn it. It is a fascinating project that is mind-warping in what it does, but it requires an investment of a lot of brainpower.

Whoever does carry the torch will come out the other side a changed developer with super-powers though.

**Comment by Skarlso** [comment]
@cowboyd Billy Mays here with another fantastic offer.... :-D never the less I'm very intrigued, but completely lacking time. I have a newborn on my hand and I have no idea how much time this would eat up. I'm guessing a lot? What do you recon? Maybe I can squeeze it in somewhere.

**Comment by Skarlso** [comment]
@cowboyd I might be able to do it with a buddy of mine who is very good at this stuff. :) I've been going over the PR, damn man, that's a LOT of stuff.

**Comment by cowboyd** [comment]
Yup, it's almost a complete re-write of the internals, so it's a lot lot of work. Like a lot lot. For those finding this later, here it is https://github.com/cowboyd/therubyracer/pull/348

It is a challenge for sure, but like all mountains worth climbing, the view from the summit is stunning.

True Story: therubyracer was conceived in my mind during the wee hours of wakefulness when I sat up with my first born back in 2009 and work began on it about 2 months later. Could it be fate?

**Comment by Skarlso** [comment]
That is quiet the awesome coincidence man. ☺️

**Comment by Skarlso** [comment]
Okay, let's say we are in. How do you want to do this? Where do we start.
